pkg/controller/usersignup/usersignup_controller_test.go

```diff
@@ -2175,3 +2176,83 @@ func TestMigrateMur(t *testing.T) {
 		assert.Equal(t, *expectedMur, murs.Items[0])
 	})
 }
+
+func TestUpdateMetricsByState(t *testing.T) {
+	t.Run("common scenarios", func(t *testing.T) {
+		metrics.Reset()
+		t.Run("empty -> not-ready - increment UserSignupUniqueTotal", func(t *testing.T) {
+			updateMetricsByState("", "not-ready")
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupAutoDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupBannedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupProvisionedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupUniqueTotal)
+		})
+
+		t.Run("not-ready -> pending - no change", func(t *testing.T) {
+			updateMetricsByState("not-ready", "pending")
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupAutoDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupBannedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupProvisionedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupUniqueTotal)
+		})
+
+		t.Run("pending -> approved - increment UserSignupProvisionedTotal", func(t *testing.T) {
+			updateMetricsByState("pending", "approved")
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupAutoDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupBannedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupProvisionedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupUniqueTotal)
+		})
+
+		t.Run("approved -> deactivated - increment UserSignupDeactivatedTotal", func(t *testing.T) {
+			updateMetricsByState("approved", "deactivated")
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupAutoDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupBannedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupProvisionedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupUniqueTotal)
+		})
+
+		t.Run("deactivated -> banned - increment UserSignupBannedTotal", func(t *testing.T) {
+			updateMetricsByState("deactivated", "banned")
+			AssertMetricsCounterEquals(t, 0, metrics.UserSignupAutoDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupBannedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupDeactivatedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupProvisionedTotal)
+			AssertMetricsCounterEquals(t, 1, metrics.UserSignupUniqueTotal)
+		})
```

Comment:
Good point about making them independent. Done. Can you see if this is what you meant?

>In addition, it would be also good to add a few checks to the unit tests that verify the whole reconcile logic - doesn't have to be everywhere, but at least a few would be good.
I went through most of the other tests and added metrics checks for whatever I thought made sense

Reply:
Yeah, it looks good now. You didn't have to put it everywhere, but it's fine as it is. Thanks a lot for it :+1: